lint reasons (RFC 2883, part 1)

zackmdavis opened the pull request:
This implements the `reason =` functionality described in [the RFC](https://github.com/rust-lang/rfcs/blob/master/text/2383-lint-reasons.md) under a `lint_reasons` feature gate.

![lint_reasons_pt_1](https://user-images.githubusercontent.com/1076988/46252097-eed51000-c418-11e8-8212-939d3f02f95d.png)

rust-highfive commented:
r? @petrochenkov

(rust_highfive has picked a reviewer for you, use r? to override)

petrochenkov commented on src/librustc/lint/levels.rs:
```diff
@@ -215,17 +214,45 @@ impl<'a> LintLevelsBuilder<'a> {
+                        ast::MetaItemKind::NameValue(ref name_value) => {
+                            let name_ident = item.ident.segments[0].ident;
+                            let name = name_ident.name.as_str();
+                            if name == "reason" {
```
IIRC, you can just use `item.ident == "reason"`.

petrochenkov commented on src/test/ui/lint/reasons-erroneous.rs:
```diff
@@ -0,0 +1,16 @@
+//~^ ERROR malformed lint attribute
+#![warn(elided_lifetimes_in_paths, reason("disrespectful to ancestors", "irresponsible to heirs"))]
+//~^ ERROR malformed lint attribute
+#![warn(ellipsis_inclusive_range_patterns, reason)]
```
Could you add a couple more cases:
`#[warn(lint1, reason = "zzz", reason = "yyy")]` - duplicated reason (probably an error).
`#[warn(lint1, reason = "zzz", lint2)]` - reason is not the last (an error? just to be conservative)

zackmdavis replied:
> reason is not the last (an error? just to be conservative)

I think this is compliant with the spirit of the RFC? @Centril @myrrlyn

This is also nice from an implementer's perspective—in the first revision of this PR, I felt bad about doing a entire extra iteration over the meta items just to pick up the reason before entering the loop that processes the lint names themselves, but with this restriction, we can just peek at the end.

petrochenkov commented on src/librustc/lint/levels.rs:
```diff
@@ -245,11 +246,33 @@ impl<'a> LintLevelsBuilder<'a> {
                                 } else {
                                     let mut err = bad_attr(name_value.span);
-                                    err.help("reason must be a string literal");
+                                    if let ast::LitKind::ByteStr(_) = name_value.node {
```
No code is good, code is evil.

petrochenkov replied:
I don't think this case will ever be hit realistically, so I see this as an immediate technical debt.

TimNN commented:
Ping from triage @zackmdavis. It looks like some changes have been requested to your PR.

Is this PR blocked on #54926?

zackmdavis commented:
@TimNN No, I just suffer from a psychological defect in which composing shiny [new PRs](https://github.com/rust-lang/rust/pull/54884) is more fun than addressing reviews of outstanding PRs (and the microincentives of which tasks are more fun play a disproportionate role in determining what humans will actually accomplish in general, not just as an open-source volunteer); give me a few more days

zackmdavis commented:
(My local build is [mysteriously failing](https://gist.github.com/zackmdavis/40837ba624133ceea5c2b57c7be730fa) right now.)

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/440512141) ([raw log](https://api.travis-ci.org/v3/job/440512141/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:20:02]    Compiling unwind v0.0.0 (/checkout/src/libunwind)
[00:20:09]    Compiling compiler_builtins v0.0.0 (/checkout/src/rustc/compiler_builtins_shim)
[00:20:09]    Compiling cmake v0.1.33
[00:20:09]    Compiling alloc_jemalloc v0.0.0 (/checkout/src/liballoc_jemalloc)
[00:20:10] error: unused import: `prelude::v1::*`
[00:20:10]     |
[00:20:10]     |
[00:20:10] 135 | use prelude::v1::*;
[00:20:10]     |
[00:20:10]     = note: `-D unused-imports` implied by `-D warnings`
[00:20:10] 
[00:20:10] error: unused macro definition
[00:20:10] error: unused macro definition
[00:20:10]   --> libcore/num/wrapping.rs:16:1
[00:20:10]    |
[00:20:10] 16 | / macro_rules! sh_impl_signed {
[00:20:10] 17 | |     ($t:ident, $f:ident) => (
[00:20:10] 18 | |         #[stable(feature = "rust1", since = "1.0.0")]
[00:20:10] 19 | |         impl Shl<$f> for Wrapping<$t> {
[00:20:10] 63 | |     )
[00:20:10] 64 | | }
[00:20:10]    | |_^
[00:20:10]    |
[00:20:10]    |
[00:20:10]    = note: `-D unused-macros` implied by `-D warnings`
[00:20:10] 
[00:20:10] error: unused macro definition
[00:20:10]    --> libcore/ops/arith.rs:636:1
[00:20:10]     |
[00:20:10] 636 | / macro_rules! neg_impl_unsigned {
[00:20:10] 637 | |     ($($t:ty)*) => {
[00:20:10] 638 | |         neg_impl_core!{ x => {
[00:20:10] 639 | |             !x.wrapping_add(1)
[00:20:10] 640 | |         }, $($t)*} }
[00:20:10]     | |_^
[00:20:10] 
[00:20:10] error: unused macro definition
[00:20:10]    --> libcore/lib.rs:237:1
[00:20:10]    --> libcore/lib.rs:237:1
[00:20:10]     |
[00:20:10] 237 | macro_rules! test_v16 { ($item:item) => {}; }
[00:20:10] 
[00:20:10] error: unused macro definition
[00:20:10]    --> libcore/lib.rs:239:1
[00:20:10]     |
[00:20:10]     |
[00:20:10] 239 | macro_rules! test_v32 { ($item:item) => {}; }
[00:20:10] 
[00:20:10] error: unused macro definition
[00:20:10]    --> libcore/lib.rs:241:1
[00:20:10]     |
[00:20:10]     |
[00:20:10] 241 | macro_rules! test_v64 { ($item:item) => {}; }
[00:20:10] 
[00:20:10] error: unused macro definition
[00:20:10]    --> libcore/lib.rs:243:1
[00:20:10]     |
[00:20:10]     |
[00:20:10] 243 | macro_rules! test_v128 { ($item:item) => {}; }
[00:20:10] 
[00:20:10] error: unused macro definition
[00:20:10]    --> libcore/lib.rs:245:1
[00:20:10]     |
[00:20:10]     |
[00:20:10] 245 | macro_rules! test_v256 { ($item:item) => {}; }
[00:20:10] 
[00:20:10] error: unused macro definition
[00:20:10]    --> libcore/lib.rs:247:1
[00:20:10]     |
[00:20:10]     |
[00:20:10] 247 | macro_rules! test_v512 { ($item:item) => {}; }
[00:20:10] 
[00:20:10] error: unused macro definition
[00:20:10]    --> libcore/lib.rs:249:1
[00:20:10]     |
[00:20:10]     |
[00:20:10] 249 | macro_rules! vector_impl { ($([$f:ident, $($args:tt)*]),*) => { $($f!($($args)*);)* } }
[00:20:10] 
[00:20:10] 
[00:20:12] error: use of deprecated item 'str::LinesAny': use lines()/Lines instead now
[00:20:12]      |
[00:20:12]      |
[00:20:12] 1387 | impl<'a> Iterator for LinesAny<'a> {
[00:20:12]      |
[00:20:12]      = note: `-D deprecated` implied by `-D warnings`
[00:20:12] 
[00:20:12] 
[00:20:12] error: use of deprecated item 'str::LinesAny': use lines()/Lines instead now
[00:20:12]      |
[00:20:12]      |
[00:20:12] 1403 | impl<'a> DoubleEndedIterator for LinesAny<'a> {
[00:20:12] 
[00:20:12] 
[00:20:12] error: use of deprecated item 'str::LinesAny': use lines()/Lines instead now
[00:20:12]      |
[00:20:12]      |
[00:20:12] 1412 | impl FusedIterator for LinesAny<'_> {}
[00:20:12] 
[00:20:12] 
[00:20:12] error: use of deprecated item 'hash::sip::SipHasher': use `std::collections::hash_map::DefaultHasher` instead
[00:20:12]    --> libcore/hash/mod.rs:100:9
[00:20:12]     |
[00:20:12] 100 | pub use self::sip::SipHasher;
[00:20:12] 
[00:20:12] 
[00:20:12] error: use of deprecated item 'hash::sip::SipHasher13': use `std::collections::hash_map::DefaultHasher` instead
[00:20:12]    --> libcore/hash/mod.rs:105:9
[00:20:12]     |
[00:20:12] 105 | pub use self::sip::SipHasher13;
[00:20:12] 
[00:20:12] 
[00:20:12] error: use of deprecated item 'hash::sip::SipHasher': use `std::collections::hash_map::DefaultHasher` instead
[00:20:12]    --> libcore/hash/sip.rs:144:6
[00:20:12] 144 | impl SipHasher {
[00:20:12]     |      ^^^^^^^^^
[00:20:12] 
[00:20:12] 
[00:20:12] error: use of deprecated item 'hash::sip::SipHasher13': use `std::collections::hash_map::DefaultHasher` instead
[00:20:12]    --> libcore/hash/sip.rs:166:6
[00:20:12] 166 | impl SipHasher13 {
[00:20:12]     |      ^^^^^^^^^^^
[00:20:12] 
[00:20:12] 
[00:20:12] error: use of deprecated item 'hash::sip::SipHasher': use `std::collections::hash_map::DefaultHasher` instead
[00:20:12]    --> libcore/hash/sip.rs:252:24
[00:20:12]     |
[00:20:12] 252 | impl super::Hasher for SipHasher {
[00:20:12] 
[00:20:12] 
[00:20:12] error: use of deprecated item 'hash::sip::SipHasher13': use `std::collections::hash_map::DefaultHasher` instead
[00:20:12]    --> libcore/hash/sip.rs:265:24
[00:20:12]     |
[00:20:12] 265 | impl super::Hasher for SipHasher13 {
[00:20:12] 
[00:20:12] 
[00:20:12] error: use of deprecated item 'hash::sip::SipHasher24': use `std::collections::hash_map::DefaultHasher` instead
[00:20:12]   --> libcore/hash/sip.rs:62:22
[00:20:12]    |
[00:20:12] 62 | pub struct SipHasher(SipHasher24);
[00:20:12] 
[00:20:12] 
[00:20:12] error: use of deprecated item 'str::LinesAny': use lines()/Lines instead now
[00:20:12]      |
[00:20:12]      |
[00:20:12] 2852 |     pub fn lines_any(&self) -> LinesAny {
[00:20:12] 
[00:20:12] 
[00:20:12] error: use of deprecated item 'str::LinesAny': use lines()/Lines instead now
[00:20:12]      |
[00:20:12]      |
[00:20:12] 2853 |         LinesAny(self.lines())
[00:20:12] 
[00:20:12] 
[00:20:12] error: use of deprecated item 'hash::sip::SipHasher': use `std::collections::hash_map::DefaultHasher` instead
[00:20:12]    --> libcore/hash/sip.rs:150:21
[00:20:12]     |
[00:20:12] 150 |     pub fn new() -> SipHasher {
[00:20:12] 
[00:20:12] 
[00:20:12] error: use of deprecated item 'hash::sip::SipHasher': use `std::collections::hash_map::DefaultHasher` instead
[00:20:12]    --> libcore/hash/sip.rs:151:9
[00:20:12]     |
[00:20:12] 151 |         SipHasher::new_with_keys(0, 0)
[00:20:12] 
[00:20:12] 
[00:20:12] error: use of deprecated item 'hash::sip::SipHasher': use `std::collections::hash_map::DefaultHasher` instead
[00:20:12]    --> libcore/hash/sip.rs:159:51
[00:20:12]     |
[00:20:12] 159 |     pub fn new_with_keys(key0: u64, key1: u64) -> SipHasher {
[00:20:12] 
[00:20:12] 
[00:20:12] error: use of deprecated item 'hash::sip::SipHasher': use `std::collections::hash_map::DefaultHasher` instead
[00:20:12]    --> libcore/hash/sip.rs:160:9
[00:20:12]     |
[00:20:12] 160 |         SipHasher(SipHasher24 {
[00:20:12] 
[00:20:12] 
[00:20:12] error: use of deprecated item 'hash::sip::SipHasher24': use `std::collections::hash_map::DefaultHasher` instead
[00:20:12]    --> libcore/hash/sip.rs:160:19
[00:20:12]     |
[00:20:12] 160 |         SipHasher(SipHasher24 {
[00:20:12] 
[00:20:12] 
[00:20:12] error: use of deprecated item 'hash::sip::SipHasher13': use `std::collections::hash_map::DefaultHasher` instead
[00:20:12]    --> libcore/hash/sip.rs:172:21
[00:20:12]     |
[00:20:12] 172 |     pub fn new() -> SipHasher13 {
[00:20:12] 
[00:20:12] 
[00:20:12] error: use of deprecated item 'hash::sip::SipHasher13': use `std::collections::hash_map::DefaultHasher` instead
[00:20:12]    --> libcore/hash/sip.rs:173:9
[00:20:12]     |
[00:20:12] 173 |         SipHasher13::new_with_keys(0, 0)
[00:20:12] 
[00:20:12] 
[00:20:12] error: use of deprecated item 'hash::sip::SipHasher13': use `std::collections::hash_map::DefaultHasher` instead
[00:20:12]    --> libcore/hash/sip.rs:181:51
[00:20:12]     |
[00:20:12] 181 |     pub fn new_with_keys(key0: u64, key1: u64) -> SipHasher13 {
[00:20:12] 
[00:20:12] 
[00:20:12] error: use of deprecated item 'hash::sip::SipHasher13': use `std::collections::hash_map::DefaultHasher` instead
[00:20:12]    --> libcore/hash/sip.rs:182:9
[00:20:12] 182 |         SipHasher13 {
[00:20:12]     |         ^^^^^^^^^^^
[00:20:12] 
[00:20:12]    Compiling std v0.0.0 (/checkout/src/libstd)
[00:20:12]    Compiling std v0.0.0 (/checkout/src/libstd)
[00:20:12] error: use of deprecated item 'str::LinesAny': use lines()/Lines instead now
[00:20:12] 
[00:20:12] error: use of deprecated item 'hash::sip::SipHasher13': use `std::collections::hash_map::DefaultHasher` instead
[00:20:12] 
[00:20:12] error: use of deprecated item 'hash::sip::SipHasher24': use `std::collections::hash_map::DefaultHasher` instead
[00:20:12] 
[00:20:12] error: use of deprecated item 'hash::sip::SipHasher': use `std::collections::hash_map::DefaultHasher` instead
[00:20:14]    Compiling rustc_msan v0.0.0 (/checkout/src/librustc_msan)
[00:20:14]    Compiling rustc_lsan v0.0.0 (/checkout/src/librustc_lsan)
[00:20:14]    Compiling rustc_tsan v0.0.0 (/checkout/src/librustc_tsan)
[00:20:15]    Compiling rustc_asan v0.0.0 (/checkout/src/librustc_asan)
[00:20:15]    Compiling rustc_asan v0.0.0 (/checkout/src/librustc_asan)
[00:20:25] error: use of deprecated item 'hash::sip::SipHasher::new_with_keys': use `std::collections::hash_map::DefaultHasher` instead
[00:20:25]    --> libcore/hash/sip.rs:151:9
[00:20:25]     |
[00:20:25] 151 |         SipHasher::new_with_keys(0, 0)
[00:20:25] 
[00:20:25] 
[00:20:25] error: use of deprecated item 'hash::sip::SipHasher24::hasher': use `std::collections::hash_map::DefaultHasher` instead
[00:20:25]    --> libcore/hash/sip.rs:161:13
[00:20:25]     |
[00:20:25] 161 |             hasher: Hasher::new_with_keys(key0, key1)
[00:20:25] 
[00:20:25] 
[00:20:25] error: use of deprecated item 'hash::sip::SipHasher13::new_with_keys': use `std::collections::hash_map::DefaultHasher` instead
[00:20:25]    --> libcore/hash/sip.rs:173:9
[00:20:25]     |
[00:20:25] 173 |         SipHasher13::new_with_keys(0, 0)
[00:20:25] 
[00:20:25] 
[00:20:25] error: use of deprecated item 'hash::sip::SipHasher13::hasher': use `std::collections::hash_map::DefaultHasher` instead
[00:20:25]    --> libcore/hash/sip.rs:183:13
[00:20:25]     |
[00:20:25] 183 |             hasher: Hasher::new_with_keys(key0, key1)
[00:20:25] 
[00:20:25] 
[00:20:25] error: use of deprecated item 'hash::sip::SipHasher24::hasher': use `std::collections::hash_map::DefaultHasher` instead
[00:20:25]    --> libcore/hash/sip.rs:255:9
[00:20:25]     |
[00:20:25] 255 |         self.0.hasher.write(msg)
[00:20:25] 
[00:20:25] 
[00:20:25] error: use of deprecated item 'hash::sip::SipHasher24::hasher': use `std::collections::hash_map::DefaultHasher` instead
[00:20:25]    --> libcore/hash/sip.rs:260:9
[00:20:25]     |
[00:20:25] 260 |         self.0.hasher.finish()
[00:20:25] 
[00:20:25] 
[00:20:25] error: use of deprecated item 'hash::sip::SipHasher13::hasher': use `std::collections::hash_map::DefaultHasher` instead
[00:20:25]    --> libcore/hash/sip.rs:268:9
[00:20:25]     |
[00:20:25] 268 |         self.hasher.write(msg)
[00:20:25] 
[00:20:25] 
[00:20:25] error: use of deprecated item 'hash::sip::SipHasher13::hasher': use `std::collections::hash_map::DefaultHasher` instead
[00:20:25]    --> libcore/hash/sip.rs:273:9
[00:20:25] 273 |         self.hasher.finish()
[00:20:25]     |         ^^^^^^^^^^^
[00:20:25] 
/modules/compiler-rt/objects
---
travis_time:end:09fe1344:start=1539331646700284145,finish=1539331646705078475,duration=4794330
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:0668a818
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:09e83770
travis_time:start:09e83770
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-g
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bors commented:
:umbrella: The latest upstream changes (presumably #55015) made this pull request unmergeable. Please resolve the merge conflicts.

zackmdavis commented:
@petrochenkov updated :checkered_flag:

petrochenkov commented on src/librustc/lint/levels.rs:
```diff
@@ -210,60 +211,70 @@ impl<'a> LintLevelsBuilder<'a> {
-                                    } else {
-                                        reason = Some(rationale);
-                                    }
+            let tail_li = &metas[metas.len()-1];
```
This will panic on `#[allow()]`, no?

zackmdavis replied:
... right. :cold_sweat: :disappointed_relieved: :skull:

zackmdavis replied:
Should this trigger `unused_attributes`? (As recently discussed for [empty derives](https://github.com/rust-lang/rust/issues/54651#issuecomment-425565263) and [inappropriate `#[must_use]`](https://github.com/rust-lang/rust/issues/54828#issuecomment-427246850))

Centril replied:
Feels reasonable to me to trigger unused_attributes for empty allow; esp. if we do it for empty derives.

petrochenkov commented on src/test/ui/lint/reasons-erroneous.rs:
```diff
@@ -12,7 +12,13 @@
+#![warn(keyword_idents, reason = "root in rubble", macro_use_extern_crate)]
+//~^ ERROR malformed lint attribute
+//~| HELP reason in lint attribute must come last
+#![warn(missing_copy_implementations, reason)]
```
What happens with `#[allow(reason = "foo")]`?
I.e. no lints to allow, only the reason.

zackmdavis replied:
Yeah, let's make this an error (it would have been a no-op if you hadn't pointed this out).

zackmdavis replied:
... _well_, actually, I guess I could see a case that, if empty lint attributes (`#[allow()]` as discussed above) aren't an error (the stable behavior), then we should also allow reason-only lint attributes for consistency/uniformity.

(Presumably the only sane reason for empty lint attributes to exist is for the sake of the base case of some recursive macro, and if we care about that, then we should also care about a macro that's otherwise identical but also includes a reason.)

zackmdavis commented:
I [took a shot](https://github.com/zackmdavis/rust/commit/0c6af84b8e0201cebea8c95586a21d97645f03e0) at implementing unused-attributes linting for empty and reason-only lint attributes, but I didn't understand some of the behavior I was seeing (the _first_ empty lint attribute in a program was getting linted twice for some reason), so I've elected to defer that, filing an issue (#55112) and leaving FIXME comments. (I could have hacked around the duplication with [one-time-diagnostics](https://github.com/rust-lang/rust/blob/df0d6ad9c1510630b234f6bb96a69a081de0b307/src/librustc/session/mod.rs#L416-L418), but that would be bad; if you want to write correct software and you don't understand something, don't _guess_.)

In this revision, we don't panic on `#[level()]` (which is already a no-op on the stable compiler), and `#[level(reason =  "foo")]` is, with some reluctance, allowed (for the [reasons mentioned above](https://github.com/rust-lang/rust/pull/54683#discussion_r225384747)). A run-pass UI test is added to document these.

... I could be persuaded to go with my first instinct and make `#[level(reason = "foo")]` an error. The "it's a generalization of `#[level()]` and needs to have the same behavior" argument seems strong to me, but another sort of conservatism argues in favor of not accepting _new_ crazy code (even as backwards-compatibility forces our hand to accept _old_ kinds of crazy code).

@petrochenkov what do you think??

petrochenkov commented:
LGTM as is.
@bors r+

bors commented:
:pushpin: Commit fecc001019a0f8d226f2596557a8eb1f8210b134 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov fecc001019a0f8d226f2596557a8eb1f8210b134 -->

bors commented:
:hourglass: Testing commit fecc001019a0f8d226f2596557a8eb1f8210b134 with merge 29a77648e840ea3e9b73a17aa383b2dc3ec327ed...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/builds/19543407)

zackmdavis commented:
Glancing at the [AppVeyor log](https://ci.appveyor.com/project/rust-lang/rust/builds/19543407/job/a25hey2cq3gg24q2), I don't see any obvious actual failure? I think I remember there being a 3-hour time limit ("Failed [...] ago in 3 hr 1 min")?

@bors retry

bors commented:
:hourglass: Testing commit fecc001019a0f8d226f2596557a8eb1f8210b134 with merge e05d425648ac44717c364a31ed61794d9b1ead43...

zackmdavis commented:
```
error: linking with `armv7-unknown-linux-gnueabihf-gcc` failed: exit code: 1
```

What? Roll to disbelieve :slightly_frowning_face: :female_detective: 

@bors retry

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/442523083?utm_source=github_status&utm_medium=notification)

rust-highfive commented:
The job `dist-x86_64-linux-alt` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/442523086) ([raw log](https://api.travis-ci.org/v3/job/442523086/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:07:11] make: *** [prepare] Error 1
[00:07:12] Command failed. Attempt 2/5:
[00:07:12]     Finished dev [unoptimized] target(s) in 0.26s
[00:07:12]  Downloading quine-mc_cluskey v0.2.4
[00:07:52] warning: spurious network error (2 tries remaining): failed to download from `https://crates.io/api/v1/crates/quine-mc_cluskey/0.2.4/download`
[00:08:33] warning: spurious network error (1 tries remaining): failed to download from `https://crates.io/api/v1/crates/quine-mc_cluskey/0.2.4/download`
[00:09:13] 
[00:09:13] Caused by:
[00:09:13] Caused by:
[00:09:13]   failed to download from `https://crates.io/api/v1/crates/quine-mc_cluskey/0.2.4/download`
[00:09:13] Caused by:
[00:09:13]   [28] Timeout was reached (Resolving timed out after 30000 milliseconds)
[00:09:13] thread 'main' panicked at 'command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "metadata" "--format-version" "1" "--features" "panic-unwind jemalloc backtrace profiler" "--manifest-path" "/checkout/src/libstd/Cargo.toml"
[00:09:13] expected success, got: exit code: 101', build_helper/lib.rs:123:9
[00:09:13] expected success, got: exit code: 101', build_helper/lib.rs:123:9
[00:09:13] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[00:09:13] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap build nonexistent/path/to/trigger/cargo/metadata
[00:09:13] Build completed unsuccessfully in 0:02:00
[00:09:13] make: *** [prepare] Error 1
[00:09:15] Command failed. Attempt 3/5:
[00:09:15]     Finished dev [unoptimized] target(s) in 0.26s
[00:09:15]  Downloading phf v0.7.22
[00:09:55] warning: spurious network error (2 tries remaining): failed to download from `https://crates.io/api/v1/crates/phf/0.7.22/download`
[00:10:35] warning: spurious network error (1 tries remaining): failed to download from `https://crates.io/api/v1/crates/phf/0.7.22/download`
[00:11:15] 
[00:11:15] Caused by:
[00:11:15] Caused by:
[00:11:15]   failed to download from `https://crates.io/api/v1/crates/phf/0.7.22/download`
[00:11:15] Caused by:
[00:11:15]   [28] Timeout was reached (Resolving timed out after 30000 milliseconds)
[00:11:15] thread 'main' panicked at 'command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "metadata" "--format-version" "1" "--features" "panic-unwind jemalloc backtrace profiler" "--manifest-path" "/checkout/src/libstd/Cargo.toml"
[00:11:15] expected success, got: exit code: 101', build_helper/lib.rs:123:9
---
travis_time:end:046f9570:start=1539758489663193393,finish=1539758489672344123,duration=9150730
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:10cf2080
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:09fefa5c
travis_time:start:09fefa5c
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:07458edc
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

zackmdavis commented:
```

[00:07:52] warning: spurious network error (2 tries remaining): failed to download from `https://crates.io/api/v1/crates/quine-mc_cluskey/0.2.4/download`
[00:08:33] warning: spurious network error (1 tries remaining): failed to download from `https://crates.io/api/v1/crates/quine-mc_cluskey/0.2.4/download`
[00:09:13] 
[00:09:13] Caused by:
[00:09:13] Caused by:
[00:09:13]   failed to download from `https://crates.io/api/v1/crates/quine-mc_cluskey/0.2.4/download`
[00:09:13] Caused by:
[00:09:13]   [28] Timeout was reached (Resolving timed out after 30000 milliseconds)
```

:crying_cat_face: :construction_worker_woman: :game_die: :skull: 

@bors retry

bors commented:
:hourglass: Testing commit fecc001019a0f8d226f2596557a8eb1f8210b134 with merge 41a0d7079906f5f8b95464018f89d37a74266c9f...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/442774090?utm_source=github_status&utm_medium=notification)

petrochenkov commented:
@bors retry

bors commented:
:hourglass: Testing commit fecc001019a0f8d226f2596557a8eb1f8210b134 with merge 1fbbc04b9279a3cdd4441661b325d677d5900cb5...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/builds/19596976)

kennytm commented:
Failed a `cargo fix` test https://github.com/rust-lang/cargo/blob/35cd245ee8522f28db62a65bfba6f8d32972d699/tests/testsuite/fix.rs#L54. 

```
[01:53:47] ---- fix::broken_fixes_backed_out stdout ----
[01:53:47] running `C:\projects\rust\build\x86_64-pc-windows-msvc\stage2-tools\x86_64-pc-windows-msvc\release\cargo.exe build`
[01:53:47] running `C:\projects\rust\build\x86_64-pc-windows-msvc\stage2-tools\x86_64-pc-windows-msvc\release\cargo.exe fix --allow-no-vcs`
[01:53:47] thread 'fix::broken_fixes_backed_out' panicked at '
[01:53:47] Expected: execs
[01:53:47]     but: expected to find:
[01:53:47] warning: failed to automatically apply fixes suggested by rustc to crate `bar`
[01:53:47] 
[01:53:47] after fixes were automatically applied the compiler reported errors within these files:
[01:53:47] 
[01:53:47]   * src/lib.rs
[01:53:47] 
[01:53:47] This likely indicates a bug in either rustc or cargo itself,
[01:53:47] and we would appreciate a bug report! You're likely to see 
[01:53:47] a number of compiler warnings after this message which cargo
[01:53:47] attempted to fix but failed. If you could open an issue at
[01:53:47] https://github.com/rust-lang/cargo/issues
[01:53:47] quoting the full output of this command we'd be very appreciative!
[01:53:47] 
[01:53:47] did not find in output:
[01:53:47]    Compiling bar v0.1.0 (C:\projects\rust\build\x86_64-pc-windows-msvc\stage2-tools\x86_64-pc-windows-msvc\cit\t563\foo\bar)
[01:53:47] error: expected one of `!` or `::`, found `rust`
[01:53:47]  --> src\lib.rs:1:5
[01:53:47]   |
[01:53:47] 1 | not rust code
[01:53:47]   |     ^^^^ expected one of `!` or `::` here
[01:53:47] 
[01:53:47] error: aborting due to previous error
[01:53:47] 
[01:53:47] error: Could not compile `bar`.
[01:53:47] warning: build failed, waiting for other jobs to finish...
[01:53:47]       Fixing src\lib.rs (1 fix)
[01:53:47] error: build failed
[01:53:47] ', tools\cargo\tests\testsuite\support\mod.rs:741:13
[01:53:47] note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

bors commented:
:umbrella: The latest upstream changes (presumably #54658) made this pull request unmergeable. Please resolve the merge conflicts.

zackmdavis commented:
> Failed a cargo fix test

@kennytm I can't reproduce this (Ubuntu 16.04). (At least not consistently—I remember seeing it _once_ while trying to reproduce the other week, while also struggling with a flaky build, but it's passing for me now, as illustrated by output below.) Please advise?

```
$ RUSTC=/home/zmd/Code/rust/build/x86_64-unknown-linux-gnu/stage2/bin/rustc cargo test broken_fixes_backed_out
   Compiling atty v0.2.11
   Compiling env_logger v0.5.13
   Compiling clap v2.32.0
   Compiling cargo v0.32.0 (file:///home/zmd/Code/cargo)
    Finished dev [unoptimized + debuginfo] target(s) in 20.96s
     Running target/debug/deps/cargo-7423faeb2a614ee4

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 23 filtered out

     Running target/debug/deps/testsuite-76f90d822272ea40

running 1 test
test fix::broken_fixes_backed_out ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 1456 filtered out
```

kennytm commented:
@zackmdavis maybe try to test on windows?

zackmdavis commented:
... I don't have a Windows machine handy. I would be pretty surprised if there was really a Windows-specific bug to find here (what Windows-specific API could this patch be using??), as opposed to something about the cargo fix test being spuriously nondeterministic (which would be _bad_, but not something we can fix in _this_ pull request). Is it OK if we try one more time?

@bors r=petrochenkov

(feel free to r- if you disagree with my judgement)

bors commented:
:pushpin: Commit f66ea66acd1c5e24e16fafe28021af1b723d6824 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov f66ea66acd1c5e24e16fafe28021af1b723d6824 -->

bors commented:
:hourglass: Testing commit f66ea66acd1c5e24e16fafe28021af1b723d6824 with merge 18311a6c47899853071cf90e37782d398dffaeed...

8573 commented:
Could the RFC number in the ticket title be corrected for ease of searching?

Also, is it right that this remains labelled as "S-waiting-on-bors"?

bjorn3 commented:
Edited the title. As for S-waiting-on-bors, it seems bors keeps it on every merged PR.
